File: 2021/day05.cpp
#include <vector>

#include "../common/puzzle.hpp"

using namespace std;
class Solution : public Puzzle {
   public:
    Solution(std::string inputFileName = "inputs/2021/5.txt")
        : Puzzle(inputFileName) {}
    std::string part1() {
        int x1, x2, y1, y2;
        vector<vector<int>> diagram(1000, vector<int>(1000));
        for (int i = 0; i < 500; i++) {
            inputFile >> x1;
            inputFile.ignore(1);
            inputFile >> y1;
            inputFile.ignore(4);
            inputFile >> x2;
            inputFile.ignore(1);
            inputFile >> y2;

            if (!(x1 == x2 || y1 == y2)) continue;
            if (y1 == y2) {
                if (x1 > x2) {
                    int tmp = x1;
                    x1 = x2;
                    x2 = tmp;
                }
                for (int j = x1; j <= x2; j++) diagram[j][y1]++;
            }
            if (x1 == x2) {
                if (y1 > y2) {
                    int tmp = y1;
                    y1 = y2;
                    y2 = tmp;
                }
                for (int j = y1; j <= y2; j++) diagram[x1][j]++;
            }
        }
        int count = 0;
        for (auto &&row : diagram)
            for (auto &&i : row)
                if (i > 1) count++;
        return std::to_string(count);
    }
    std::string part2() {
        int x1, x2, y1, y2;
        vector<vector<int>> diagram(1000, vector<int>(1000));
        for (int i = 0; i < 500; i++) {
            inputFile >> x1;
            inputFile.ignore(1);
            inputFile >> y1;
            inputFile.ignore(4);
            inputFile >> x2;
            inputFile.ignore(1);
            inputFile >> y2;

            if (x1 == x2 || y1 == y2) {
                if (y1 == y2) {
                    if (x1 > x2) {
                        int tmp = x1;
                        x1 = x2;
                        x2 = tmp;
                    }
                    for (int j = x1; j <= x2; j++) diagram[j][y1]++;
                } else if (x1 == x2) {
                    if (y1 > y2) {
                        int tmp = y1;
                        y1 = y2;
                        y2 = tmp;
                    }
                    for (int j = y1; j <= y2; j++) diagram[x1][j]++;
                }
            } else if (x1 - y1 == x2 - y2) {
                if (x1 > x2) {
                    int tmp = x1;
                    x1 = x2;
                    x2 = tmp;
                    tmp = y1;
                    y1 = y2;
                    y2 = tmp;
                }
                // cout << x1 << "," << y1 << "->" << x2 << "," << y2 << "\n";
                for (int j = 0; j <= x2 - x1; j++) {
                    diagram[x1 + j][y1 + j]++;
                    // cout << x1 + j << " " << y1 + j << "\n";
                }
            } else if (x1 + y1 == x2 + y2) {
                if (x1 > x2) {
                    int tmp = x1;
                    x1 = x2;
                    x2 = tmp;
                    tmp = y1;
                    y1 = y2;
                    y2 = tmp;
                }
                // cout << x1 << "," << y1 << "->" << x2 << "," << y2 << "\n";
                for (int j = 0; j <= x2 - x1; j++) {
                    diagram[x1 + j][y1 - j]++;
                    // cout << x1 + j << " " << y1 - j << "\n";
                }
            }
        }
        int count = 0;
        for (auto &&row : diagram)
            for (auto &&i : row)
                if (i > 1) count++;
        return std::to_string(count);
    }
};

int main() {
    Solution s = Solution();
    s.solve();
}

File: year2016/day05.java
package year2016;

import java.security.*;
import java.util.Scanner;
import java.math.BigInteger;

import common.puzzle;

public class day05 extends puzzle {
    public day05() {
        super("./inputs/2016/5.txt");
    }

    public static void main(String[] args) {
        new day05().solve();
    }

    protected String part1() {
        Scanner reader = getScanner();
        String doorId = reader.nextLine();
        String result = "";
        try {
            int index = 0;
            for (int i = 0; i < 8; i++) {
                String hashtext = "";
                while (!hashtext.startsWith("00000")) {
                    MessageDigest m = MessageDigest.getInstance("MD5");
                    m.reset();
                    m.update((doorId + Integer.toString(index++)).getBytes());
                    byte[] digest = m.digest();
                    BigInteger bigInt = new BigInteger(1, digest);
                    hashtext = bigInt.toString(16);
                    while (hashtext.length() < 32)
                        hashtext = "0" + hashtext;
                }
                result += hashtext.charAt(5);
            }
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        return result;
    }

    protected String part2() {
        Scanner reader = getScanner();
        String doorId = reader.nextLine();
        StringBuilder result = new StringBuilder("________");
        try {
            int index = 0;
            while (result.toString().contains("_")) {
                String hashtext = "";
                while (!hashtext.startsWith("00000")) {
                    MessageDigest m = MessageDigest.getInstance("MD5");
                    m.reset();
                    m.update((doorId + Integer.toString(index++)).getBytes());
                    byte[] digest = m.digest();
                    BigInteger bigInt = new BigInteger(1, digest);
                    hashtext = bigInt.toString(16);
                    while (hashtext.length() < 32)
                        hashtext = "0" + hashtext;
                }
                int passIndex = (int) (hashtext.charAt(5) - '0');
                if (passIndex >= 0 && passIndex <= 7 && result.charAt(passIndex) == '_')
                    result.setCharAt(passIndex, hashtext.charAt(6));
            }
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        return result.toString();
    }
}